waveparser.go

```diff
@@ -0,0 +1,164 @@
+		DataBlockSize  uint32 // size of sample block (PCM data)
+	}
+)
+
+func (hdr *WavHeader) String() string {
+	strs := []string{
+		"=== RIFF Header ===",
+		fmt.Sprintf("RIFF Ident: %s", string(hdr.RIFFHdr.Ident[:])),
+		fmt.Sprintf("RIFF Size: %d bytes", hdr.RIFFHdr.ChunkSize),
+		fmt.Sprintf("File type: %s", string(hdr.RIFFHdr.FileType[:])),
+		"=== Fmt ===",
+		fmt.Sprintf("Audio format: %d", hdr.RIFFChunkFmt.AudioFormat),
+		fmt.Sprintf("Number of channels: %d", hdr.RIFFChunkFmt.NumChannels),
+		fmt.Sprintf("Sample rate: %d", hdr.RIFFChunkFmt.SampleRate),
+		fmt.Sprintf("Bytes/seconds: %d", hdr.RIFFChunkFmt.BytesPerSec),
+		fmt.Sprintf("Bytes/block: %d", hdr.RIFFChunkFmt.BytesPerBloc),
+		fmt.Sprintf("Bits/sample: %d", hdr.RIFFChunkFmt.BitsPerSample),
+	}
+	return strings.Join(strs, "\n")
+}
+
+func parseRIFFHeader(r io.Reader) (*riffHeader, error) {
+	var hdr riffHeader
+	err := binary.Read(r, binary.LittleEndian, &hdr)
+	if err != nil {
+		return nil, err
+	}
+	if string(hdr.Ident[:]) != "RIFF" {
+		return nil, fmt.Errorf("Invalid RIFF identification: %s", string(hdr.Ident[:]))
+	}
+	return &hdr, nil
+}
+
+func parseHeader(r io.ReadSeeker) (*WavHeader, error) {
+	riffhdr, err := parseRIFFHeader(r)
+	if err != nil {
+		return nil, err
+	}
+
+	// FMT chunk
+	var chunk [4]byte
+	var chunkFmt riffChunkFmt
+
+	err = binary.Read(r, binary.LittleEndian, &chunk)
+	if err != nil {
+		return nil, err
+	}
+	if string(chunk[:]) != "fmt " {
+		return nil, fmt.Errorf("Unexpected chunk type: %s", string(chunk[:]))
+	}
+
+	err = binary.Read(r, binary.LittleEndian, &chunkFmt)
+	if err != nil {
+		return nil, err
+	}
+
+	if chunkFmt.AudioFormat != 1 {
+		return nil, fmt.Errorf("Isn't an audio format: format[%d]", chunkFmt.AudioFormat)
+	}
+
+	if chunkFmt.LengthOfHeader != 16 {
+		var extraparams uint16
+		// Get extra params size
+		if err = binary.Read(r, binary.LittleEndian, &extraparams); err != nil {
+			return nil, fmt.Errorf("error getting extra fmt params: %s", err)
+		}
+		// Skip
+		if _, err = r.Seek(int64(extraparams), os.SEEK_CUR); err != nil {
+			return nil, fmt.Errorf("error skipping extra params: %s", err)
+		}
+	}
+
+	var chunkSize uint32
+
+	for string(chunk[:]) != "data" {
+		// Read chunkID
+		err = binary.Read(r, binary.BigEndian, &chunk)
+		if err != nil {
+			return nil, fmt.Errorf("Expected data chunkid: %s", err)
+		}
+
+		err = binary.Read(r, binary.LittleEndian, &chunkSize)
+		if err != nil {
+			return nil, fmt.Errorf("Expected data chunkSize: %s", err)
+		}
+
+		// ignores LIST chunkIDs (unused for now)
+		if string(chunk[:]) != "data" {
+			if _, err = r.Seek(int64(chunkSize), os.SEEK_CUR); err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	pos, _ := r.Seek(0, os.SEEK_CUR)
+	return &WavHeader{
+		RIFFHdr:      *riffhdr,
+		RIFFChunkFmt: chunkFmt,
+
+		FirstSamplePos: uint32(pos),
+		DataBlockSize:  uint32(chunkSize),
+	}, nil
+}
+
+func LoadAudio(audiofile string) (*WavHeader, []int16, error) {
```

Comment:
Not that returning just two values is a rule...but usually to me it is some concept missing. Here it seems like it would be a Wav struct, like this:

```go
type Wav struct {
    Header WavHeader
    Audio []int16
}
```

It maps better to how the original audio file is represented (a header + audio contents).

Reply:
I agree. But maybe we can leave it as a future improvement. Even because right now there are places where I am using this function and ignoring the Header, so it's kind of useful to do: `_, audio, err := LoadAudio("audio.wav")`.